docs/github-app.adoc

```diff
@@ -0,0 +1,140 @@
+= GitHub app authentication guide
+
+This guide is targeted to users who want to use a link:https://developer.github.com/v3/apps/[GitHub app]
+to authenticate to Jenkins.
+
+== Why?
+
+- the link:https://developer.github.com/apps/building-github-apps/understanding-rate-limits-for-github-apps/[rate limit]
+for a GitHub app scales with your organization size, whereas a user based token has a limit of 5000 regardless of
+how many repositories you have,
+- for organization's that have 2fa enforced - no need to manage 2fa tokens for a 'bot' user
+- to improve and tighten security: the Jenkins GitHub app requires a minimum, controlled set of privileges compared to a service user and its personal access token which has a much wider set of privileges
+
+== Getting started
+
+Before you get started make sure you have the required permissions:
+
+=== GitHub
+
+You'll need the permission to create a GitHub app, if you're creating it on a personal account then you can skip this section,
+otherwise:
+
+- organization owner
+
+or
+
+- permission to manage GitHub apps has been
+link:https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/adding-github-app-managers-in-your-organization[delegated to you].
+
+=== Jenkins
+
+You'll need the permission to create a new credential and update job configuration, the specific permissions are:
+
+- Credentials/Create
+- Job/Configure
+
+== Creating the GitHub app
+
+link:https://developer.github.com/apps/building-github-apps/creating-a-github-app/[Follow the GitHub guide for creating an app]
+
+The only fields you need to fill out (currently) are:
+
+- Github App name - i.e. `Jenkins - <team name>`
+- Homepage URL - your company domain or a github repository
+- Webhook URL - your jenkins instance, i.e. `https://<jenkins-host>/github-webhook/`
```

Comment:
Is this actually used? AFAICT GH does not send anything to this URL except a notification that the app has been created, even if you select other events in the app configuration; and if you want to actually trigger builds after PRs are opened and the like, you still need to configure a webhook in the org/repo settings (or allow Jenkins to do this on your behalf via `GitHubOrgWebHook.register`) just like you would when using a PAT for authentication.

Reply:
Yeah you are correct with the fact that github actually sends out all events to that webhook, when pull requests are created, check-suites are created etc all payloads are delivered via that webhook.I am not quite familiar with the github branch source code, but if at all the payloads need capturing this webhook proves to be quite useful

Reply:
not used as far as I know, at the time I couldn't see a way to disable it, but there seems to be a checkbox now (maybe I missed it earlier), so maybe this isn't needed

Reply:
To correct myself: it suffices to configure the app to receive _Push_ and _Pull request_ events. These will be sent to the app’s webhook address and `github-branch-source` will process them. You do not need to configure a regular webhook on the repository. This means that `GitHubOrgWebHook` is useless in this case.

Reply:
⇒ #289